### PR TITLE
[doc] Better document for Draft-Target-Model (DTM) speculative decoding

### PR DESCRIPTION
**Pull request by wili-65535**
## Description

Previous [PR3366](https://github.com/NVIDIA/TensorRT-LLM/pull/3366) about DTM document, cherry-pick from [PR3655](https://github.com/NVIDIA/TensorRT-LLM/pull/3655).

A more detailed document about using Draft-Target-Model in Triton Inference Server.

Target of this PR:

+ Add steps to use symmetrical or asymmetrical TP size for draft and target engine.
+ More detailed steps to use Fast-Logits-D2D.
+ Steps to combine the two features above.
+ Note that Streaming mode or Batched-Request mode are not supported till the current version.

### CODE REVIEW

File: examples/draft_target_model/README.md [hunk 1]
```diff
@@ -224,17 +228,19 @@ python3 tools/inflight_batcher_llm/speculative_decoding_test.py \
...
 
-7. Stop triton inference server after all work is done
++ Stop triton inference server after all work is done.
 
 ```bash
 pkill tritonserver
 ```
 
-8. Advanced usage: Fast logits D2D transfer.
++ In addition, it appears better performance if both draft and target engine are deployed in one GPU (llama-7B-FP8 + llama-30B-FP8, totally 40GiB in one H100-80GiB GPU as example) than in two separate GPUs.
```
Comment (achartier) [inline examples/draft_target_model/README.md]:
```suggestion
+ In addition, it appears better performance can be achieved with both draft and target engines deployed on a single GPU (llama-7B-FP8 + llama-30B-FP8, for a total of 40GiB on one H100-80GiB GPU for example).
```

File: examples/draft_target_model/README.md [hunk 2]
```diff
@@ -224,17 +228,19 @@ python3 tools/inflight_batcher_llm/speculative_decoding_test.py \
...
 
-+ Fast logits boosts the performance (TPS) by hiding the latency of logits transfer from draft engine to target engine supported since TensorRT-LLM-0.15.0.
+#### Usage of Tensor-Parallelization mode.
 
-+ Modify `participant_ids` entry in `tensorrt_llm/config.pbtxt` and `tensorrt_llm_draft/config.pbtxt` to suitable MPI ranks. Usually in this setting, rank 0 is reserved for the orchestrator rank; rank 1 is for draft engine; the rest of the ranks are for target engine. In this example, `particpant_ids` can be set as snippet below. Same logic also applies to TP>1 target engine.
++ For example, we have built draft engine with TP=1 and target engine with TP=2 (both symmetrical or asymmetrical TP size are acceptable), and want to place the draft engine on GPU0, target engine on GPU1.
```
Comment (achartier) [inline examples/draft_target_model/README.md]: This looks odd, target engine should need 2 GPUs since TP=2?

File: examples/draft_target_model/README.md [hunk 3]
```diff
@@ -224,17 +228,19 @@ python3 tools/inflight_batcher_llm/speculative_decoding_test.py \
...
++ Edit model configuration the same as step `3.`, where still set `DRAFT_DEVICE_IDS="0"` and ``TARGET_DEVICE_IDS="1"` since comma is not supported in script `python3 tools/fill_template.py`.
++ Fix the device IDs in `${TRITON_MODEL_REPO}/tensorrt_llm/config.pbtxt` and `${TRITON_MODEL_REPO}/tensorrt_llm/config.pbtxt` manually as:
```
Comment (achartier) [inline examples/draft_target_model/README.md]: sed can be used for that purpose:
```
sed -i 's/\${gpu_device_ids}/0/g' ${TRITON_MODEL_REPO}/tensorrt_llm_draft/config.pbtxt
sed -i 's/\${gpu_device_ids}/1,2/g' ${TRITON_MODEL_REPO}/tensorrt_llm/config.pbtxt
```

File: examples/draft_target_model/README.md [hunk 4]
```diff
@@ -244,64 +250,99 @@ parameters: {
...
     value: {
-        string_value: "1"
+        string_value: "1,2"
     }
 }
-### In tensorrt_llm/config.pbtxt
+```
+
++ Start the triton inference server
+  + `--multi-model` must be specified if asymmetrical TP is used.
```
Comment (achartier) [inline examples/draft_target_model/README.md]: I thought it was required regardless of TP size?

File: examples/draft_target_model/README.md [hunk 5]
```diff
@@ -244,64 +250,99 @@ parameters: {
...
+        string_value: "2,3"
     }
 }
 ```
 
-+ Launched Triton Server
-  + Use orchestrator mode with `--disable-spawn-process`. See [model config](https://github.com/triton-inference-server/tensorrtllm_backend/blob/main/docs/model_config.md) for more information.
-  + `--world_size` has to be set as 1 (orchestrator rank 0) + 1 (draft engine ranks) + 1 (target engine ranks).
++ Start the triton inference server.
+  + `--disable-spawn-process` to enable orchestrator mode. See [model config](https://github.com/triton-inference-server/tensorrtllm_backend/blob/main/docs/model_config.md) for more information.
```
Comment (achartier) [inline examples/draft_target_model/README.md]: `--multi-model` is enabling orchestrator mode. `--disable-spawn-process` is setting the pre-spawn variant of orchestrator mode.

File: examples/draft_target_model/README.md [hunk 6]
```diff
@@ -244,64 +250,99 @@ parameters: {
...
+  + `--world_size` must be equal to (tp_draft+tp_target+1). It is 4 in our example.
```
Comment (achartier) [inline examples/draft_target_model/README.md]: Might be worth documenting the +1 is for the orchestrator rank (if not documented elsewhere)

File: examples/draft_target_model/README.md [hunk 7]
```diff
@@ -244,64 +250,99 @@ parameters: {
...
-    -H "Content-Type: application/json" \
-    -d '{
-        "text_input": "Continue writing the following story: James Best, best known for his",
-        "max_tokens": 128,
-        "num_draft_tokens": 10,
-        "use_draft_logits": true,
-        "stream": false
-        }'
-```
++ All other operations is the same as `Simple deploy` part.
```
Comment (achartier) [inline examples/draft_target_model/README.md]:
```suggestion
+ All other operations are the same as the `Simple deploy` part.
```

File: examples/draft_target_model/README.md (hunk 2 above)
Comment (wili-65535) [reply examples/draft_target_model/README.md]: Sorry for the typo, it should be "target engine on GPU1 and GPU2".
I fixed it as (https://github.com/NVIDIA/TensorRT-LLM/pull/3797/files#diff-cb8abaf07b6038374f6a384f472c42ba849459670132997b26d691025e156c01R242).

File: examples/draft_target_model/README.md (hunk 3 above)
Comment (wili-65535) [reply examples/draft_target_model/README.md]: Thanks for the suggestions, I add the commands to replace `gpu_device_ids` and `participant_ids` as [Line262](https://github.com/NVIDIA/TensorRT-LLM/pull/3797/files#diff-cb8abaf07b6038374f6a384f472c42ba849459670132997b26d691025e156c01R262) and [Line307](https://github.com/NVIDIA/TensorRT-LLM/pull/3797/files#diff-cb8abaf07b6038374f6a384f472c42ba849459670132997b26d691025e156c01R307). It makes the steps simple and elegant.

File: examples/draft_target_model/README.md (hunk 4 above)
Comment (wili-65535) [reply examples/draft_target_model/README.md]: I found it worked smoothly in symmetrical TP mode without `--multi-model`, but failed in asymmetrical TP mode.
Anyway, we unify the usage as [Line269](https://github.com/NVIDIA/TensorRT-LLM/pull/3797/files#diff-cb8abaf07b6038374f6a384f472c42ba849459670132997b26d691025e156c01R269), using `--multi-model` in any TP>1 scenario.

File: examples/draft_target_model/README.md (hunk 5 above)
Comment (wili-65535) [reply examples/draft_target_model/README.md]: Thanks for the explanation about those parameters. I adjust related text as [Line269](https://github.com/NVIDIA/TensorRT-LLM/pull/3797/files#diff-cb8abaf07b6038374f6a384f472c42ba849459670132997b26d691025e156c01R269) and [Line316](https://github.com/NVIDIA/TensorRT-LLM/pull/3797/files#diff-cb8abaf07b6038374f6a384f472c42ba849459670132997b26d691025e156c01R316).

File: examples/draft_target_model/README.md (hunk 6 above)
Comment (wili-65535) [reply examples/draft_target_model/README.md]: We have noted it in previous version [Line237](https://github.com/NVIDIA/TensorRT-LLM/pull/3797/files#diff-cb8abaf07b6038374f6a384f472c42ba849459670132997b26d691025e156c01L237)
But for emphasis, we move it here [Line284](https://github.com/NVIDIA/TensorRT-LLM/pull/3797/files#diff-cb8abaf07b6038374f6a384f472c42ba849459670132997b26d691025e156c01R284).

File: examples/draft_target_model/README.md (hunk 7 above)
Comment (wili-65535) [reply examples/draft_target_model/README.md:328]: Fixed!

**Review by litaotju — APPROVED** [review]

**Comment by litaotju** [comment]
Approving for the change, this is fine to submit to 0.19.

@wili-65535 pls let me know to merge it after you addressed all the comments from @achartier . Thx

File: examples/draft_target_model/README.md (hunk 7 above)
Comment (achartier) [reply examples/draft_target_model/README.md:328]: L278 is fixed, but L328 still has "is the same", unless I'm misreading github diff?

Comment (wili-65535) [reply examples/draft_target_model/README.md]: No, that 's my fault of incorrect git rebase operation, now it is OK [Line328](https://github.com/NVIDIA/TensorRT-LLM/pull/3797/files#diff-cb8abaf07b6038374f6a384f472c42ba849459670132997b26d691025e156c01R328).

**Review by achartier — APPROVED** [review]
LGTM

**Comment by wili-65535** [comment]
Thanks a lot @achartier , @litaotju !
